Http probe request object reuse

Ahmed-Elgamel opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind cleanup
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Avoid creating a new request object everytime when making an http probe instead the worker  just constructs the request once and every http probe uses the same request object
#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->
Fixes #115939

#### Special notes for your reviewer:

I am open to any modification to my approach for request reuse.

My implemntation was as follows:
1) created a new object that will store the request object
2) The worker when making an http probe will create a new request (if none was created) or use the same request that was created earlier
3) For all the probing fucntions, I modified them to accept  a new argument, which is the request object ---it can be null (old behavior) or have a value that was provided by the worker (the reused request object)

If there’s a better approach or anything you’d like me to change, I’m happy to do so.


#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

Ahmed-Elgamel committed:
ADD new probe object that will store the http request

Ahmed-Elgamel committed:
ADD the object that stores the http request probe inside the worker struct

Ahmed-Elgamel committed:
CHANGE all probing fuctions to take a request parameter as input

Ahmed-Elgamel committed:
PROBING WORKER can now call the new probing fuction and pass in the same request without everytime constructing a new request object

pacoxu commented:
/cc @aojea @SergeyKanzhelev 
per your comments https://github.com/kubernetes/kubernetes/issues/115939#issuecomment-3094823643

haircommander commented on pkg/kubelet/prober/worker.go:
```diff
@@ -288,8 +321,22 @@ func (w *worker) doProbe(ctx context.Context) (keepGoing bool) {
-	result, err := w.probeManager.prober.probe(ctx, w.probeType, w.pod, status, w.container, w.containerID)
+	var result results.Result
+	var err error
+	// Use cached HTTP request if available
```
I think this could be simplified if you do 
```
var req *http.Request
if w.httpProbeRequest != nil {
		req, reqErr := w.httpProbeRequest.getRequest(&w.container) // either returns cached request or creates a new one
		if reqErr != nil {
			klog.V(4).InfoS("HTTP-Probe failed to create request", "error", reqErr)
			return true
		}
}
		// Note, exec probe does NOT have access to pod environment variables or downward API
		result, err = w.probeManager.prober.probe(ctx, w.probeType, w.pod, status, w.container, w.containerID, req)
```
i.e. drop the else and duplicatred probe call.

Ahmed-Elgamel replied:
updated the code to simplify as you suggested 👍

Ahmed-Elgamel committed:
REFACTOR probe call

haircommander commented on pkg/kubelet/prober/prober.go:
```diff
@@ -155,7 +156,19 @@ func (pb *prober) runProbe(ctx context.Context, probeType probeType, p *v1.Probe
 	case p.HTTPGet != nil:
+		if req != nil {
+			// Use the provided request object
+			if klogV4 := klog.V(4); klogV4.Enabled() {
```
I don't understand this. shouldn't you just do 
```
klog.V(4).InfoS("HTTP-Probe using cached request",
					"scheme", req.URL.Scheme, "host", req.URL.Hostname(),
					"port", req.URL.Port(), "path", req.URL.Path)
```
and it would acheive the same?

SergeyKanzhelev commented on pkg/kubelet/prober/prober.go:
```diff
@@ -155,7 +156,19 @@ func (pb *prober) runProbe(ctx context.Context, probeType probeType, p *v1.Probe
+		if req != nil {
+			// Use the provided request object
+			if klogV4 := klog.V(4); klogV4.Enabled() {
+				klogV4.InfoS("HTTP-Probe using cached request",
```
you do not really need this new log every time we probe something.

Ahmed-Elgamel replied:
I initially thought of making request reuse as an optional behavior, which is why I assumed req could be nil and added a fallback to the old behavior. However, I now understand that the intended direction is to always reuse the request object, so I’ll proceed with that and only log the probe when the request is created and not when reused.

SergeyKanzhelev reviewed (commented):
this is new approach that introduces the least amount of changes at once. Many previous attempts were a lot of refactoring at once.

I like this, even though I believe more refactoring is needed in this package.

Please keep working on it - add some unit tests and clean up logging. We should come back to it very early in 1.35, once we done with 1.34

Ahmed-Elgamel committed:
kubelet: simplify and refactor HTTP probe request handling in prober
- Remove the fallback HTTP request creation logic in the prober file
- Remove the request object null check since the http probing fuction (runProbe) will not be called if the worker failed to create the http request object and the worker will keep on trying again
- Improve error logging

Ahmed-Elgamel committed:
Add HTTP probe request holder initialization in newWorker

Initialize httpProbeRequestHolder for HTTP probes in newWorker function
The holder is created with empty podIP initially, which will be
set dynamically during probe execution (in worker loop).

This approach handles cases where:
- Pod is created but not yet scheduled to a node (no IP available)
- Pod IP changes during pod lifecycle (rescheduling, network changes)
- Pod IP becomes available after worker creation

The podIP is intentionally left empty in newWorker because:
1. Pod may not be scheduled to a node yet when worker is created
2. Pod IP can change during pod lifecycle
3. Dynamic IP assignment in worker loop ensures current IP is always used (next step)

Next step: Handle pod IP changes in worker loop by updating/resetting
the httpProbeRequestHolder when IP changes are detected.

Ahmed-Elgamel committed:
Add pod IP change handling for HTTP probe requests

Reset/Invalidate cached HTTP request when pod IP changes to ensure probes
always use the current pod IP address. This handles pod rescheduling
and network configuration changes during the pod lifecycle.

Ahmed-Elgamel committed:
ADD unit test for probe worker http request caching

Ahmed-Elgamel commented:
I added a benchmark function cad75c477730e676c7f089233feae23c8df6f7ff to evaluate this approach for optimizing HTTP probes. The benchmark follows the same structure from previous bencmharks for this issue.

```
goos: darwin
goarch: arm64
pkg: k8s.io/kubernetes/pkg/kubelet/prober
cpu: Apple M2
                        │ nocaching.txt │    caching.txt     │
                        │    sec/op     │   sec/op     vs base   │
WorkerProbe_NoCaching-8     21.17m ± 3%
WorkerProbe_Caching-8                     250.8µ ± 2%
geomean                     21.17m        250.8µ       ? ¹ ²
¹ benchmark set differs from baseline; geomeans may not be comparable
² ratios must be >0 to compute geomean

                        │ nocaching.txt │     caching.txt     │
                        │     B/op      │     B/op      vs base   │
WorkerProbe_NoCaching-8    74.69Mi ± 0%
WorkerProbe_Caching-8                     385.8Ki ± 0%
geomean                    74.69Mi        385.8Ki       ? ¹ ²
¹ benchmark set differs from baseline; geomeans may not be comparable
² ratios must be >0 to compute geomean

                        │ nocaching.txt │    caching.txt     │
                        │   allocs/op   │  allocs/op   vs base   │
WorkerProbe_NoCaching-8     859.1k ± 0%
WorkerProbe_Caching-8                     3.634k ± 0%
geomean                     859.1k        3.634k       ? ¹ ²
¹ benchmark set differs from baseline; geomeans may not be comparable
² ratios must be >0 to compute geomean
```

Ahmed-Elgamel commented:
/auto-cc

matthyx commented:
/ok-to-test

Ahmed-Elgamel committed:
REFACTOR probe worker test file

Ahmed-Elgamel commented:
@SergeyKanzhelev @haircommander will this pr be merged or is there review to be done on it still?

Ahmed-Elgamel commented:
/test pull-kubernetes-unit-windows-master

SergeyKanzhelev commented:
rebase, squash commits. And I also hoped for some review with lgtm on this. We are close to code freeze again, so likely this will miss the deadline again
